### DISCUSSION TITLE
Android engine/engines for TTS and STT/ASR

### DISCUSSION THREAD

**Discussion by paolo-caroni** [k2-fsa/sherpa-onnx#1392] [Ideas]
I am writing here because I think the topic is more suitable for discussions than "issues". As requested by several people (see #852 #569 #1088 #569 ), sherpa-onnx would need an engine app for TTS (and a separate or merged one for STT/ASR #580 ). Android offers a standard [API](https://developer.android.com/reference/android/speech/tts/TextToSpeech) for TTS that is currently partially implemented by android .apk packages that support one language at a time and that do not allow the use of multilingual apps (or change language easy).
To save space on devices and implement the choice of language, the engine app should be a single one and allow the language models to be installed separately (without being already present in the package).

An attempt to start the work in this direction was started by @Jing332 at the [repository](https://github.com/jing332/SherpaOnnxTtsEngineAndroid) but the development stopped at the initial stage.

I would like to understand how I can help in the development of a stable and working app that can reach the level of integration at least equal to eSpeak (I would hope even higher).
Could a fundraiser be useful?

As for the STT/ASR engine the situation is worse. Currently the [API](https://developer.android.com/reference/kotlin/android/speech/SpeechRecognizer) is not supported at all. Language detection support [DETECTED_LANGUAGE](https://developer.android.com/reference/kotlin/android/speech/SpeechRecognizer#DETECTED_LANGUAGE:kotlin.String) is possible, as it is currently done by a [spoken language identification](https://k2-fsa.github.io/sherpa/onnx/spoken-language-identification/apk.html) apk that doesn't seem to interface with the API.
Sherpa-onnx seems to have all the capabilities to support language detection, multi-language support, confidence level and more.

Regarding merging or keeping separate TTS and STT, the models are different, so for the same language you would need a model for TTS and a model for STT, so it would probably be easier to have separate engines.

As pointed out [here](https://github.com/k2-fsa/sherpa-onnx/issues/1088#issuecomment-2216713532) the purpose of the android packages does not seem to be to offer an engine that supports the native android API, but rather to provide examples on how to directly integrate sherpa-onnx in each app with a different API. This would require, if I understood correctly, that each app that listens or spoke to the user should integrate sherpa-onnx, which for obvious reasons makes no sense on Android platform.

Again, I would not be able to write useful code, but I could think about participating in a crowdfunding.

You who have been interested in the topic previously, what do you think?
@mablue @Jing332 @csukuangfj @thewh1teagle @domasofan @gyroing

**Comment by mablue** [comment]
WhT you think about it?
I think the fastest way to reach to the more goals is hide behind tts-server:
https://github.com/jing332/tts-server-android/discussions/222

**Reply by paolo-caroni** [reply]
I agree that probably can be fast, but only for one goal, the tts implementation. Honestly I think that try to correct the buggy implementation of an tts app that download the language, will be helpful for a stt app in the future.
Honestly I hope that someone will take care of it in the near future. Also tts-server seems to.need different apk for each language, correct? A different apk for each language will lead to various problem on market/store.

**Reply by mablue** [reply]
**I disagree that tts-server needs a separate apk for each language.** It actually manages a list of online and offline languages internally. Similar to Jing332's multi-Sherpa engine, I believe we can easily download additional models like Sherpa ONNX zip files from the internet or storage.

By incorporating these downloaded models, we can implement speech-to-text (STT) and automatic speech recognition (ASR) functionalities for writing rules. Additionally, we could potentially merge all rule types (reading, writing, replace) into a single "Rules" tab for better workflow.

The reason I chose tts-server is because of its extensive features, including reading rules and plugins. It offers similar functionalities to Jieshuo ([https://github.com/nirenr/jieshuo](https://github.com/nirenr/jieshuo)), but it's not open-source, had potential trojans in older versions, and isn't free. Many blind Iranians use cracked versions riddled with malware due to the lack of alternatives. 

In other countries like Turkey, blind users rely on this screen reader for its speed and plugin functionalities. Similarly, deaf individuals lack accessible offline tools for communication. By integrating Sherpa functionality into tts-server, we can create a powerful tool for people with various disabilities.

For languages like Pashto, Microsoft Read Aloud offers limited support with two additional voices. tts-server, however, could potentially provide online access to this functionality until a dedicated Pashto Sherpa model becomes available. Unfortunately, the lack of big data hinders Icefall's development in this area.

Viewing tts-server as a multilingual accessibility tool with Sherpa model support can potentially address many current issues, including routine problems and limited language support.




Current Challenges:
 * Limited Language Support: Existing TTS engines lack offline support for languages like Pashto.
 * Accessibility Issues: Many accessibility tools for the blind are closed-source, non-free, or contain malware. Deaf communities lack offline accessibility tools.
 * Fragmented Functionality: TTS servers often require separate apps for each language and lack integrated functionalities like Sherpa for writing assistance.

Proposed Solution:
 * Extend TTS Server Functionality: We propose enhancing an existing TTS server like tts-server (similar to jieshuo) to incorporate Sherpa functionalities for writing assistance. This would create a powerful, multilingual accessibility tool.
 * Offline Model Download: Implement the ability to download and manage offline language models (like Sherpa's ONNX models) directly within the TTS server, eliminating the need for multiple apps. Models could be downloaded from the internet or local storage.
 * Unified Rule Management: Combine reading, writing, and replacement rules under a single "Rules" tab for a more streamlined user experience.
 
Benefits:
 * Improved Accessibility: This solution would provide a secure, offline accessibility tool for both blind and deaf communities, catering to a wider range of disabilities.
 * Multilingual Support: The ability to download and manage offline language models would address limitations in languages like Pashto and offer greater language flexibility.
 * Enhanced Functionality: Integrating Sherpa functionalities would provide additional writing assistance tools, improving overall user experience.
 
Addressing Concerns:
 * Security: We acknowledge the security risks associated with cracked versions of jaeshuo+ screenreader. By promoting a secure, open-source alternative, we can ensure user safety.
 * Data Availability: While a Pashto Sherpa model may not be readily available, this solution paves the way for future integration when such models become accessible.
 
End Note:
 * By viewing tts-server as a core accessibility tool and enabling Sherpa model support, we can create a powerful, secure, and multilingual solution that addresses the critical needs of diverse disabled communities.
 * Each day we delay, many people are unable to read, write, hear, or participate fully in life. With every passing day, their opportunities diminish. Let's not let their voices go unheard.

**Reply by paolo-caroni** [reply]
tts server seems to me localized only in chineese, and is only a TTS, not an STT, so I don't understand why you are choosing it.
sherpa-onnx seems actually to me more mature.

Also @mablue i think you are mixing different things.

Sherpa-onnx is a program of a community (kaldi k2 community).
Ask here to implement a language like Pashto is at least little rude.
Pashto have a free dataset? Yes, on Common Voice
It is complete? Not really

so what YOU can do?
Improve the existant Pashto dataset (https://commonvoice.mozilla.org/ps)
Today 94 hour, 1206 speakers, 9154 sentences, 45% of validation progress (These values are very low).
So you can join Common Voice community (different from kaldi k2 community) and help to improve them.

You can also use icefall (made by kaldi k2 community) to add the existant Common Voice Pashto to sherpa-onnx (but since the actual status of the dataset don't expect too much, but still you can do it today if you want to help Pashto blinds).

As you can see YOU can solve one of your points (the first of your points).
Don't expect that other solve for you, in particular no one can improve Pashto without the ability of read and speach Pashto languge (if you decide to improve the dataset please read carefully the rules).

My points of view is different, my first point is how to make sherpa-onnx worldwide used but much people possible, so they can contribute and improve it.
For that I think (but is only my modest opinion) that the best way is download of the dataset directly from the TTS engine app, so that app can be add to all the existant stores (first of my list is f-droid, but also google play store, Samsung Galaxy Store, Huawei app gallery, Amazon Appstore, ecc.).
For that I cannot help in any ways, except to pay people that have the ability for do it, but for a fund raise are needed two different things:
1) people that give money (I'm alone, I cannot give enought money);
2) people that do the job and take the money.

Consider also that a STT/ASR engine app of sherpa-onnx simply don't exist yet, there are only proof of concept.

The points of view of the kaldi k2 community is still different, for them Android is only one of the supported plattform, and sadly them don't seem to want to spend too much energy and time (and I respect there point of view, like yours).

**Reply by mablue** [reply]
@paolo-caroni 
Imagine a world with your point of view: A Sherpa-Onnx TTS engine with multiple language download functionality available on F-Droid. How would this impact your life?
Consider the perspective of a stone that can accidentally think, see, move, and hear: How might this ability allow it to help other stones?
Please let me know!
What should happen to it after 80 years?  How long did it have to wait to understand this beautiful world, to understand its luck of self-consciousness?
What is it except for pure atoms?  What is its point of view on these situations?
Having a toy or maybe making money by making toys for other stones?  Or giving a gift to any other stone that accidentally falls into a disability?
We need to chain together to end these disabilities... We need each other.
Thank you, it's very helpful resources for me. I will try to make a Pashto model.  We also have a multi-Sherpa model created by jing332 that works very well. Here's the link: https://github.com/jing332/SherpaOnnxTtsEngineAndroid

**Reply by paolo-caroni** [reply]
> Or giving a gift to any other stone that accidentally falls into a disability? We need to chain together to end these disabilities... We need each other.

Wait a moment, this is exactly the concept of free software. I have just pointed to you that you cannot ask for a gift. You can make a gift or pay someone to make a gift to others. You cannot ask to the other stone to rumble when you stand still.
If I have sounded aggressive I am very sorry, I don't whant to give you any offense.
Please consider other peoples view.

> Thank you, it's very helpful resources for me. I will try to make a Pashto model.

Your welcome, I'm improving my mother language on Common Voice since 4 years.
Pashto need surely people that cooperate on it.

> We also have a multi-Sherpa model created by jing332 that works very well. Here's the link: https://github.com/jing332/SherpaOnnxTtsEngineAndroid

I know it, but on my phone it doesn't seems to work. And as you have pointed me on other issue, he now work for xiaomi so I suppose that him will stop to develop it.

**Reply by mablue** [reply]
Why it might not work on your phone:
It's interesting that the app doesn't work on your phone. There might still be some bugs for certain phone models. To help us fix it, please send us your app logs.
TTS server rule manager for switching between languages:
We've implemented a TTS server rule manager that allows switching between multiple text-to-speech engines while reading. This is particularly useful for languages like Persian, where you might encounter English words within the text. Persian TTS models can't read English words effectively. The TTS server automatically switches between Persian and English models to ensure smooth reading.
Example JS rule used in the TTS server app:
```javascript
let SpeechRuleJS = {
  name: "FesgheliX",
  id: "ir.masoudsoft.ttsfarsi.rr.fesgheliX",
  author: "Masoud Azizi",
  telegram: "@ttsfarsi",
  version: 5,
  tags: { en: "English", fa: "Farsi" },

  handleText(text) {
    // Split text based on word boundaries, identifying English words
    return text.split(/([a-zA-Z][a-zA-Z\s]+[a-zA-Z])/).map(part => ({
      text: part,
      tag: /[a-zA-Z][a-zA-Z\s]+[a-zA-Z]/.test(part) ? "en" : "fa"
    }));
  },
  splitText(text) {
    // Split text based on punctuation and delimiters
    return text.split(/([,;:?\[\]{}\"\'`„“”‚‘’؟!«»().،]+)/);
  },
};
```
Learn more about our workflow:
For a deeper understanding of the app's workflow and our team's work, please visit our channel:
https://t.me/addlist/8JNmLa_MTJAwNzZk

**Reply by paolo-caroni** [reply]
I undestand this needs, but your proposed solution is not standard.
This will brake compatibility if the user use other tts engine. A similar function in my honest opinion should be implemented on the app that send the text to the engine, not on the engine itself. So for the blind will be implemented on talkback? Also, a persian model with some known popular english word should solve it in a different way. However I will look more about tts engine. I have tested it founding only chineese text on the options, chineese as only supported languege so I give up. I will retry it.

**Reply by mablue** [reply]
Well, partner, this here Google Accessibility Tools is supposed to be the bee's knees, but it's slower'n a turtle on cold molasses. Especially when we're trying to rope in two language models with that multi sherpa and tts-server rigamarole.
Now, those screen readers like jieshou+ and prodence ain't half bad, but they're a lot of work to set up. It'd be a heap of a lot better if we could just have one app that does it all, like a one-stop shop for accessibility.
And let's face it, a screen reader and a text-to-speech engine are just about the same thing, if ya ask me. It's like a cowboy and his trusty steed. One can't do much without the other. So why not just have one app that can handle both? That way, we could use it with any text-to-speech engine we want, just like that tts-server.
In Android, we got two different ways to do it. One's for the default screen reader, and the other's straight for TalkBack. And guess what? The second one don't care if the screen reader's even turned on or not.

**Reply by paolo-caroni** [reply]
Your code assume that if there is an alphabetical character the word is english. But I think this is wrong, can't be spanish? or italian? or french? The best way to implement this shouldn't be a language identification? ML Kit is closed source, but fastext-language-identification is open source for example.
This imho should be implemented on the screen reader app, not on the TTS engine.

In any case merge screen reader and TTS engine it's not prohibited, you can do what do you want, but why mantain big code if you can split up? TTS engine is installed and used on all the smartphone, by everyone, but the screen reder is used/installed only in a fraction of them. So as you can see, TTS engine and screen reader are not the same thing. Are like a horse and a cowboy, the cowboy cannot ride without a horse, but a horse can run free or with a jockey instead of a cowboy. Also, I think that the user should decide what engine use, I have tested sherpa onnx on a Huawei Sonic, it's almost unusable, for old/slow smartphone espeak-ng is still the best option.

**Reply by mablue** [reply]
i said appending language code to end of tts engine files can be usefull to people that want to have two or 3 languages of tts

**Reply by powline99** [reply]
@paolo-caroni @mablue I am sorry to interrupt your thread but I can't seem to find a solution and perhaps you can help. 
I am not advanced enough to get into the coding of this model etc... 

I am just trying to get it to work on Android for use in my Accessibility TTS readers. I see there were created apks https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html . 

When I install on my Google Pixel it only allows one to be used at a time (no input or way to change to a different voice's apk under the app)? It installs the new voice over the same APK from the download. 

Is there a easier way to do this with Fdroid app where you can just select the voice you want off the onnx voices?

Thanks,
DD.

**Reply by paolo-caroni** [reply]
> @paolo-caroni @mablue I am sorry to interrupt your thread but I can't seem to find a solution and perhaps you can help. I am not advanced enough to get into the coding of this model etc...

Hi, this is not a issue, but a discudsion, so any request for help is a bit off-topic, also there are issues about that (cited on my first post) have you read them?

> I am just trying to get it to work on Android for use in my Accessibility TTS readers. I see there were created apks https://k2-fsa.github.io/sherpa/onnx/tts/apk-engine.html .
> 
> When I install on my Google Pixel it only allows one to be used at a time (no input or way to change to a different voice's apk under the app)? It installs the new voice over the same APK from the download.
> 
> Is there a easier way to do this with Fdroid app where you can just select the voice you want off the onnx voices?

No, not yet, sadly maybe never, try the jing332 fork https://github.com/jing332/SherpaOnnxTtsEngineAndroid but for me doesn't run as expected.
Obviously it's not yet on fdroid, you have to install it directly from github.

**Reply by paolo-caroni** [reply]
> Is there a easier way to do this with Fdroid app where you can just select the voice you want off the onnx voices?
> 
There is a new fork called SherpaTTS, is in active development, you find it in fdroid.

**Comment by rocksit** [comment]
I'm not sure if it is helpful, but there are plenty a lot of similar working foss SST android applications / solutions available. I found for instance 
- FOSS SST App Sayboard https://f-droid.org/packages/com.elishaazaria.sayboard/
- K6nele SST full android system integration  https://github.com/Kaljurand/K6nele?tab=readme-ov-file based on a K6nele Service mapper https://github.com/Kaljurand/K6nele-service,  where actually a Kaldi GStreamer Server is connected https://github.com/alumae/kaldi-gstreamer-server, but it should be possible, to replace this in favour of another backend.
- Dicio https://github.com/Stypox/dicio-android is another foss SST android system integration, that tries to provide usefull speech triggerd tasks and plain SST to android.. 
Hope this could serve as inspiration.
Even if they were already mentioned, here is a further details link on jieshuo and prudence screen reader apps, that gives some good overview https://accessibleandroid.com/jieshuo-screen-reader/ and https://accessibleandroid.com/prudence-screen-reader/
